DBZ-7368: add more documentation for converters in outbox router (#5222)
debezium/debezium: documentation/modules/ROOT/pages/transformations/outbox-event-router.adoc

* {link-prefix}:{link-smt-predicates}#applying-transformations-selectively[Configure an SMT predicate for the transformation].
* Use the xref:outbox-event-router-property-route-topic-regex[`route.topic.regex`] configuration option for the SMT.

// Type: concept
// Title: Payload serialization format
// ModuleID: payload-serialization-format
== Payload serialization format

> The outbox event router SMT supports arbitrary payload formats. The `payload` column value in an outbox table is passed on transparently. However, the outbox event router SMT needs to be configured correctly to convert the data from the database column to a Kafka message (in other words, to serialize the payload). Common payload formats are JSON and Avro.

Author: It makes sense to me. This paragraph is mostly based on existing content, so I don't have any strong opinion about it.

// Type: concept
// Title: Using JSON as the serialization format
// ModuleID: using-json-payload-format
[[using-json-payload-format]]
=== Using JSON as the payload format

JSON is the default serialization format for this SMT. In order to use this format the database column must be of JSON format (i.e. `jsonb` in PostgreSQL).

// Type: concept
// Title: Producing Expanding escaped JSON String as JSON
// ModuleID: expanding-escaped-json-string-as-json
[[expanding-escaped-json-string-as-json]]
==== Expanding escaped JSON String as JSON

> You may have noticed that the Debezium outbox message contains the `payload` represented as a String.
> So when this string, is actually JSON, it appears as escaped in the result Kafka message like shown below:

Reviewer: ```suggestion
When a {prodname} outbox message represents the `payload` as a JSON String, the resulting Kafka message  escapes the string as in the following example:
```

[source,javascript,indent=0]
----
# Kafka Topic: outbox.event.order
# Kafka Message key: "1"
# Kafka Message Headers: "id=4d47e190-0402-4048-bc2c-89dd54343cdc"
# Kafka Message Timestamp: 1556890294484
{
  "{\"id\": 1, \"lineItems\": [{\"id\": 1, \"item\": \"Debezium in Action\", \"status\": \"ENTERED\", \"quantity\": 2, \"totalPrice\": 39.98}, {\"id\": 2, \"item\": \"Debezium for Dummies\", \"status\": \"ENTERED\", \"quantity\": 1, \"totalPrice\": 29.99}], \"orderDate\": \"2019-01-31T12:13:01\", \"customerId\": 123}"
}
----

> The outbox event router allows you to expand this message content to "real" JSON with the companion schema
> being deduced from the JSON document itself. That way the result in Kafka message looks like:

Reviewer: ```suggestion
The outbox event router enables you to expand the message content to "real" JSON, deducing the companion schema from the JSON document. 
The resulting Kafka message is formatted as in the following example:
```

[source,javascript,indent=0]
----
# Kafka Topic: outbox.event.order
# Kafka Message key: "1"
# Kafka Message Headers: "id=4d47e190-0402-4048-bc2c-89dd54343cdc"
# Kafka Message Timestamp: 1556890294484
{
  "id": 1, "lineItems": [{"id": 1, "item": "Debezium in Action", "status": "ENTERED", "quantity": 2, "totalPrice": 39.98}, {"id": 2, "item": "Debezium for Dummies", "status": "ENTERED", "quantity": 1, "totalPrice": 29.99}], "orderDate": "2019-01-31T12:13:01", "customerId": 123
}
----

> To enable this transformation, you have to set the xref:outbox-event-router-property-table-expand-json-payload[`table.expand.json.payload`] to true and use the `JsonConverter` like below:

Reviewer: ```suggestion
To enable use of the outbox event router transformation, set the xref:outbox-event-router-property-table-expand-json-payload[`table.expand.json.payload`] to true, and use the `JsonConverter` as shown in the following example:
```

[source]
----
transforms=outbox,...
transforms.outbox.type=io.debezium.transforms.outbox.EventRouter
transforms.outbox.table.expand.json.payload=true
value.converter=org.apache.kafka.connect.json.JsonConverter
----

// Type: concept
// Title: Using Avro as the payload format in {prodname} outbox messages

> // ModuleID: using-avro-as-the-payload-format-in-debezium-outbox-messages

Reviewer: ```suggestion
// ModuleID: outbox-event-router-using-avro-as-the-payload-format-in-debezium-outbox-messages
```

[[avro-as-payload-format]]
=== Using Avro as the payload format

> A common practice is to serialize data as Avro. This can be beneficial for message format governance and for ensuring that outbox event schemas evolve in a backwards-compatible way.

Reviewer: ```suggestion
A common practice is to serialize data as Avro. 
Using Avro can be beneficial for message format governance and for ensuring that outbox event schemas evolve in a backwards-compatible way.
```

// Type: concept
// Title: Payload serialization format
// ModuleID: outbox-event-router-payload-serialization-format
== Payload serialization format

The outbox event router SMT supports arbitrary payload formats. 
The SMT passes on `payload` column values that it reads from the outbox table without modification. 
The way that the SMT converts these column values into Kafka message fields depends on how you configure the SMT.
Common payload formats for serializing data are JSON and Avro.

// Type: concept
// Title: Using JSON as the serialization format
// ModuleID: outbox-event-router-using-json-payload-format
[[using-json-payload-format]]
=== Using JSON as the payload format

The default serialization format for the outbox event router SMT is JSON. 
To use this format, the data type of the source column must be JSON (for example, `jsonb` in PostgreSQL).

// Type: concept
// Title: Producing Expanding escaped JSON String as JSON
// ModuleID: outbox-event-router-expanding-escaped-json-string-as-json
[[expanding-escaped-json-string-as-json]]
==== Expanding escaped JSON String as JSON

> When a {prodname} outbox message represents the `payload` as a JSON String, the resulting Kafka message  escapes the string as in the following example:

Author: ```suggestion
When a {prodname} outbox message represents the `payload` as a JSON String, the resulting Kafka message escapes the string as in the following example:
```

}
----

The outbox event router enables you to expand the message content to "real" JSON, deducing the companion schema from the JSON document. 
The resulting Kafka message is formatted as in the following example:

[source,javascript,indent=0]
----
# Kafka Topic: outbox.event.order
# Kafka Message key: "1"
# Kafka Message Headers: "id=4d47e190-0402-4048-bc2c-89dd54343cdc"
# Kafka Message Timestamp: 1556890294484
{
  "id": 1, "lineItems": [{"id": 1, "item": "Debezium in Action", "status": "ENTERED", "quantity": 2, "totalPrice": 39.98}, {"id": 2, "item": "Debezium for Dummies", "status": "ENTERED", "quantity": 1, "totalPrice": 29.99}], "orderDate": "2019-01-31T12:13:01", "customerId": 123
}
----

To enable use of the outbox event router transformation, set the xref:outbox-event-router-property-table-expand-json-payload[`table.expand.json.payload`] to true, and use the `JsonConverter` as shown in the following example:

[source]
----
transforms=outbox,...
transforms.outbox.type=io.debezium.transforms.outbox.EventRouter
transforms.outbox.table.expand.json.payload=true
value.converter=org.apache.kafka.connect.json.JsonConverter
----

// Type: concept

> // Title: Using Avro as the payload format in {prodname} outbox messages

Reviewer: ```suggestion
// Title: Using Apache Avro as the payload format in {prodname} outbox messages
```

Reviewer: I know that this is pre-existing content, and not directly related to this PR, but I'd appreciate it if you could correct this oversight as part of your update. 🙏🏼

// ModuleID: outbox-event-router-using-avro-as-the-payload-format-in-debezium-outbox-messages
[[avro-as-payload-format]]

> === Using Avro as the payload format

Reviewer: ```suggestion
=== Using Apache Avro as the payload format
```

> A common practice is to serialize data as Avro. 

Reviewer: ```suggestion
Apache Avro is a common framework for serializing data. 
```

value.converter=io.debezium.converters.BinaryDataConverter
----

> By default, the `payload` column value (the Avro data) is the only message value. When storing data in Avro format the column must be of binary format (i.e. `bytea` in PostgreSQL),

Reviewer: ```suggestion
By default, the `payload` column value (the Avro data) is the only message value. 
When data is stored in Avro format, the column format must be set to a binary data type, such as `bytea` in PostgreSQL.
```

> and value converter for the SMT must be `BinaryDataConverter`, which will propagate the `payload` column binary value as-is into the Kafka message value.

Reviewer: ```suggestion
The value converter for the SMT must also be set to `BinaryDataConverter`, so that it propagates the binary value of the `payload` column as-is into the Kafka message value.
```

The delegate `Converter` implementation is specified by the `delegate.converter.type` option.
If any extra configuration options are needed by the converter, they can also be specified, such as the disablement of schemas shown above using `schemas.enable=false`.

> Furthermore, the following example illustrates how to use a delegate converter in Avro format with Schema Registry:

Reviewer: ```suggestion
The following example illustrates how to use a delegate converter in Avro format with Schema Registry:
```

Reviewer: Should the reference to _Schema Registry_ be changed to _Apicurio Registry_?

Author: I don't have access to a running instance of Apicurio Registry in order to test the setup and provide an example. But yes, if you think it would be better to document Apicurio, we can do it. Would the example below would still be valid?

Reviewer 2: Hi, yes, I'd prefer to have either Apicurio registry example or both. Please check https://github.com/debezium/debezium/blob/main/debezium-connector-postgres/pom.xml#L227 how do we start the simplest Apicurio registry deployment. It can be started easily via `docker run`

Author: Added example for apicurio

[source]
----
transforms=outbox,...
transforms.outbox.type=io.debezium.transforms.outbox.EventRouter
value.converter=io.debezium.converters.BinaryDataConverter
value.converter.delegate.converter.type=io.confluent.connect.avro.AvroConverter
value.converter.delegate.converter.type.basic.auth.credentials.source=USER_INFO
value.converter.delegate.converter.type.basic.auth.user.info={CREDENTIALS}
value.converter.delegate.converter.type.schema.registry.url={URL}
----

[NOTE]
====

> The last example with an AvroConverter as the delegate converter requires third party libraries. Adding those libraries to the classpath out of the scope for this document.

Reviewer: ```suggestion
In the preceding configuration example, because the `AvroConverter` is configured as a delegate converter, third-party libraries are required. 
Information about how to add third-party libraries to the classpath is beyond the scope of this document.
```
